File: cloud-storage-spring-java/src/main/java/com/github/sevtech/cloud/storage/spring/service/dropbox/DropboxService.java
package com.github.sevtech.cloud.storage.spring.service.dropbox;

import com.dropbox.core.DbxDownloader;
import com.dropbox.core.DbxException;
import com.dropbox.core.v2.DbxClientV2;
import com.dropbox.core.v2.files.FileMetadata;
import com.github.sevtech.cloud.storage.spring.model.DeleteFileRequest;
import com.github.sevtech.cloud.storage.spring.model.DeleteFileResponse;
import com.github.sevtech.cloud.storage.spring.model.GetFileRequest;
import com.github.sevtech.cloud.storage.spring.model.GetFileResponse;
import com.github.sevtech.cloud.storage.spring.model.UploadFileRequest;
import com.github.sevtech.cloud.storage.spring.model.UploadFileResponse;
import com.github.sevtech.cloud.storage.spring.service.AbstractStorageService;
import com.github.sevtech.cloud.storage.spring.service.StorageService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.HttpStatus;
import org.springframework.scheduling.annotation.Async;
import org.springframework.scheduling.annotation.AsyncResult;

import java.io.IOException;
import java.util.concurrent.Future;

@Slf4j
@RequiredArgsConstructor
public class DropboxService extends AbstractStorageService implements StorageService {

    private final DbxClientV2 dbxClientV2;

    @Override
    public UploadFileResponse uploadFile(final UploadFileRequest request) {
        try {
            log.info("Uploading file to {}", getFilePath(request));
            dbxClientV2.files().uploadBuilder("/" + getFilePath(request))
                    .uploadAndFinish(request.getStream());
            return UploadFileResponse.builder().fileName(request.getName()).status(HttpStatus.SC_OK).build();
        } catch (final DbxException | IOException e) {
            return UploadFileResponse.builder().fileName(request.getName()).status(HttpStatus.SC_INTERNAL_SERVER_ERROR)
                    .cause(e.getLocalizedMessage()).exception(e).build();
        }
    }

    @Async
    @Override
    public Future<UploadFileResponse> uploadFileAsync(final UploadFileRequest request) {
        return new AsyncResult<>(uploadFile(request));
    }

    @Override
    public GetFileResponse getFile(final GetFileRequest request) {
        try {
            log.info("Reading file from {}", request.getPath());
            final DbxDownloader<FileMetadata> download = dbxClientV2.files().download(request.getPath());
            return GetFileResponse.builder().content(
                    new byte[download.getInputStream().available()]).status(HttpStatus.SC_OK).build();
        } catch (final DbxException | IOException e) {
            return GetFileResponse.builder().status(HttpStatus.SC_INTERNAL_SERVER_ERROR)
                    .cause(e.getLocalizedMessage()).exception(e).build();
        }
    }

    @Override
    public DeleteFileResponse deleteFile(final DeleteFileRequest request) {
        try {
            log.info("Deleting file from path {}", request.getPath());
            dbxClientV2.files().deleteV2(request.getPath());
            return DeleteFileResponse.builder().result(true).status(HttpStatus.SC_OK).build();
        } catch (final DbxException e) {
            return DeleteFileResponse.builder().status(HttpStatus.SC_INTERNAL_SERVER_ERROR)
                    .cause(e.getLocalizedMessage()).exception(e).build();
        }
    }
}

File: cloud-storage-spring-kotlin/src/main/kotlin/com/github/sevtech/cloud/storage/spring/service/dropbox/DropboxService.kt
package com.github.sevtech.cloud.storage.spring.service.dropbox

import com.dropbox.core.DbxException
import com.dropbox.core.v2.DbxClientV2
import com.github.sevtech.cloud.storage.spring.model.*
import com.github.sevtech.cloud.storage.spring.service.AbstractStorageService
import com.github.sevtech.cloud.storage.spring.service.StorageService
import org.apache.http.HttpStatus
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Async
import org.springframework.scheduling.annotation.AsyncResult
import java.io.IOException
import java.util.concurrent.Future

class DropboxService(private val dbxClientV2: DbxClientV2) : AbstractStorageService(), StorageService {

    private val log = LoggerFactory.getLogger(this.javaClass)

    override fun uploadFile(request: UploadFileRequest): UploadFileResponse {
        return try {
            log.info("Uploading file to ${getFilePath(request)}")
            dbxClientV2.files().uploadBuilder("/" + getFilePath(request))
                    .uploadAndFinish(request.stream)
            UploadFileResponse(request.name, HttpStatus.SC_OK)
        } catch (e: DbxException) {
            UploadFileResponse(fileName = request.name, status = HttpStatus.SC_INTERNAL_SERVER_ERROR,
                    cause = e.localizedMessage, exception = e)
        } catch (e: IOException) {
            UploadFileResponse(fileName = request.name, status = HttpStatus.SC_INTERNAL_SERVER_ERROR,
                    cause = e.localizedMessage, exception = e)
        }
    }

    @Async
    override fun uploadFileAsync(request: UploadFileRequest): Future<UploadFileResponse> {
        return AsyncResult(uploadFile(request))
    }

    override fun getFile(request: GetFileRequest): GetFileResponse {
        return try {
            log.info("Reading file from {}", request.path)
            val download = dbxClientV2.files().download(request.path)
            GetFileResponse(ByteArray(download.inputStream.available()), HttpStatus.SC_OK)
        } catch (e: DbxException) {
            GetFileResponse(status = HttpStatus.SC_INTERNAL_SERVER_ERROR,
                    cause = e.localizedMessage, exception = e)
        } catch (e: IOException) {
            GetFileResponse(status = HttpStatus.SC_INTERNAL_SERVER_ERROR,
                    cause = e.localizedMessage, exception = e)
        }
    }

    override fun deleteFile(request: DeleteFileRequest): DeleteFileResponse {
        return try {
            log.info("Deleting file from {}", request.path)
            dbxClientV2.files().deleteV2(request.path)
            DeleteFileResponse(true, HttpStatus.SC_OK)
        } catch (e: DbxException) {
            DeleteFileResponse(status = HttpStatus.SC_INTERNAL_SERVER_ERROR,
                    cause = e.localizedMessage, exception = e)
        }
    }

}